### ISSUE TITLE
Empty content collection when dev is first run

### ISSUE THREAD

**Issue by ericswpark** [withastro/astro#12866]
### Astro Info

```block
Astro                    v5.1.1
Node                     v23.3.0
System                   Windows (x64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/tailwind
```

### If this issue only occurs in one browser, which browser is a problem?

N/A

### Describe the Bug

On Windows, Astro reports an empty Content Layer collection when running the server with `npm run dev`. Using the built preview (`npm run build; npm run preview`) is not affected and shows all content within the Content Layer collection. I've tested the same site on Linux and it does not suffer from this issue.

### What's the expected result?

Astro should show the proper content within the Content Layer collection, instead of showing nothing.

### Link to Minimal Reproducible Example

N/A - platform-specific bug

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by ericswpark** [comment]
This is what I get when running `npm run dev` and navigating to my blog collection at `/blog/`:

```
The collection "blog" does not exist or is empty. Ensure a collection directory with this name exists.
20:00:46 [200] /blog/ 19ms
```

Here is the definition, which is the same as what is shown on the Astro migration guide:

```
import { defineCollection, z } from "astro:content";
import { glob } from "astro/loaders";

const blogCollection = defineCollection({
  loader: glob({ pattern: "**/[^_]*.{md,mdx}", base: "./src/content/blog" }),
  schema: z.object({
    title: z.string(),
    author: z.string().optional(),
    date: z.coerce.date(),
    tags: z.array(z.string()).optional(),
  }),
});

export const collections = {
  blog: blogCollection,
};
```

I initially thought it might be a path delimiter issue (`/` vs `\`), but again, building and running a preview works fine, so Astro _does_ detect my content on Windows -- just not on the dev server.

**Comment by ericswpark** [comment]
Even though I get a `RenderUndefinedEntryError` upon trying to load one of the blog posts, this issue does _not_ seem to be related to issue #12773. I still get the error with the `experimental--dev-after-sync` branch.

**Comment by ericswpark** [comment]
@florian-lefebvre as I mentioned in the OP a repro would be of no use – Astro works fine on Linux and subsequently most of the web-based platforms that presumably is Linux container-based. This appears to be a platform specific issue.

**Comment by Gazook89** [comment]
I actually have what I believe is the same problem, but I am on Mac.  My experience with the issue is documented in the discord support channel [discord](https://discord.com/channels/830184174198718474/1324200357357944892/1324200357357944892), but in summary:

I cannot get a migration to v5 to work on a very simple blog because content collections are empty.  I initiated a new Astro Blog template project on both IDX and [Stackblitz (minimal example)](https://stackblitz.com/edit/withastro-astro-bwq4pw7d) and had no issue.  So then I downloaded the zip file from Stackblitz to my machine, uncompressed it, did `npm install` + `npm run dev` and had the issue with the collections again.

So I don't believe it's any issue with not following instructions, since it seems to work from the SB and IDX templates when used in those environments but just doesn't work locally.

Finding this thread just now, I tried `npm run build` + `npm run preview` which succeeded with both the downloaded SB template and my own personal project.

Again, I am on the current MacOS version, and had issues on both Node 20 and 22, and all Astro dependencies were up to date as far as I could tell.   

I couldn't find any other errors in browser or in VS Code.

**Comment by florian-lefebvre** [comment]
@ericswpark I understand it's OS specific but a repro is still useful for us because we can't take the time to replicate manually each issue. It doesn't have to be a stackblitz, a repo with the minimal conditions to reproduce your bug on windows is absolutely fine! We'll check it out on windows too

**Comment by florian-lefebvre** [comment]
@Gazook89 in the project downloaded locally, can you run `astro info` and paste the result here?

**Comment by ericswpark** [comment]
@florian-lefebvre understood, I thought we could only submit web container based from the issue form. Will submit a repro repo in a second!

**Comment by ericswpark** [comment]
@florian-lefebvre this is very weird and annoying. It appears that there is some sort of race condition going on here.

I had my site repo cloned to `repoA`, with the broken collections error message (see comment https://github.com/withastro/astro/issues/12866#issuecomment-2566352005). I created an empty Astro project, brought over the minimal amount of code to replicate the collection page, and it worked.

I then re-cloned my repository to a separate directory `repoA.reclone`, ran `npm i` and `npm run dev`. This one worked, so I tried to figure out what was different between `repoA` and `repoA.reclone`, but everything was identical, except for the generated `.astro/` directory and the `node_modules/` directory. I then removed `.astro` and `node_modules` from repo A, re-ran `npm i` and `npm run dev`, and now `repoA` is rendering correctly as well with no collection error.

Keep in mind that I did this several times in repo A -- deleting all generated files and dependencies and reinstalling and re-running the dev server -- and it still didn't work until today. I can try and remove the `.astro/` folder and `node_modules` and try reinstalling over and over until I can hit the race condition again.

Here is a link to the minimal example repo: https://github.com/ericswpark/astro-issue-12866-repro/
Although again, I will probably have to reinstall and rerun a bunch of times to hit the race condition again.

**Comment by ascorbic** [comment]
I'm pretty sure this is a dupe of #12773. I know you've said `experimental--dev-after-sync` doesn't fix it, but can you ensure that you're using that in the repro.

**Comment by ericswpark** [comment]
@ascorbic I already tried changing to that branch with my repository and still encountered the same issue. I don't remember if I removed `.astro/` and `node_modules` but I'm pretty sure I did do that as a troubleshooting step, just in case. As my repository is no longer in the bugged state I most likely won't be able to reproduce unless I can run into that race condition again.

Feel free to close the issue -- I will comment if I ever re-run into this problem with the changes from that branch merged.

**Comment by ericswpark** [comment]
@ascorbic I've managed to recreate the issue with the experimental branch, and believe I have proper repro instructions:

1. Remove `.astro/`
2. Install `npm i` with experimental branch set in `package.json`
3. `npm run build`
4. `npm run dev`

Step 3 appears to be important -- without step 3, `npm run dev` shows all content in the collection, and subsequent calls to `npm run build` will not trigger the bugged state.

I've tried this 4~5 times and have gotten it to reliably break on all of the attempts. Please let me know if you can reproduce this problem.

**Comment by ascorbic** [comment]
In your repro, despite the version in the package.json it's installing 5.1.1. Try doing `npm i astro@experimental--dev-after-sync` manually and see if it works for you. If that doesn't, can you share the logs from when you run `astro dev`

**Comment by ericswpark** [comment]
@ascorbic you're right, I missed that part. However, I finally figured out why the experimental branch broke for me as well. Try this:

1. Manual install with `npm i astro@experimental--dev-after-sync`
2. `npm run astro sync`
3. `npm run build`
4. `npm run dev`

Again, skipping either 2 or 3 before running 4 will result in not hitting this edge case.

Once you reproduce, your `.astro/` folder should be missing the `data-store.json` file.

After running step 4, you should get a log output like this:

```
$ npm run dev

> ericswpark.com@0.0.1 dev
> astro dev

22:24:18 [types] Generated 1ms
22:24:18 [WARN] [content] Content config not loaded
22:24:18 [vite] Re-optimizing dependencies because vite config has changed

 astro  v0.0.0-dev-after-sync-20241223135326 ready in 888 ms

┃ Local    http://localhost:4321/
┃ Network  use --host to expose

▶ This is a  dev-after-sync-20241223135326  prerelease build!
  Report issues here: https://astro.build/issues

22:24:18 watching for file changes...
The collection "blog" does not exist or is empty. Please check your content config file for errors.
22:25:44 [200] /blog/ 401ms
The collection "blog" does not exist or is empty. Please check your content config file for errors.
22:25:44 [200] HEAD /ko/blog/ 19ms
22:25:44 [200] /_image 43ms
```

**Comment by Gazook89** [comment]
I just woke up so haven't run through the whole thread here, but for @florian-lefebvre here is my `astro info`:

```
Astro                    v5.1.1
Node                     v22.12.0
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
```

**Comment by ascorbic** [comment]
@ericswpark "Content config not loaded" is the important bit there. It means Vite can't load the content config. Do you get that in the minimal repro?

**Comment by ascorbic** [comment]
@Gazook89 can you try with `npm i astro@experimental--dev-after-sync`

**Comment by Gazook89** [comment]
@ascorbic doing `npm i astro@experimental--dev-after-sync` didn't seem to work for me.  It installed, and `npm run dev` after didn't help.  I also tried doing the `build` and retrying `dev` to no effect.  

Here is my terminal from when i tried this with a basic blog template with no further changes after running the init script:

<details><summary>terminal</summary>

```
User@user withastro-astro-bwq4pw7d % npm i astro@experimental--dev-after-sync
npm warn ERESOLVE overriding peer dependency
npm warn While resolving: @example/blog@0.0.1
npm warn Found: astro@5.1.1
npm warn node_modules/astro
npm warn   peer astro@"^5.0.0" from @astrojs/mdx@4.0.3
npm warn   node_modules/@astrojs/mdx
npm warn     @astrojs/mdx@"^4.0.3" from the root project
npm warn   1 more (the root project)
npm warn
npm warn Could not resolve dependency:
npm warn peer astro@"^5.0.0" from @astrojs/mdx@4.0.3
npm warn node_modules/@astrojs/mdx
npm warn   @astrojs/mdx@"^4.0.3" from the root project

changed 1 package, and audited 361 packages in 1s

178 packages are looking for funding
  run `npm fund` for details

found 0 vulnerabilities
User@user withastro-astro-bwq4pw7d % npm run dev                             

> @example/blog@0.0.1 dev
> astro dev

08:36:07 [types] Generated 0ms
08:36:07 [content] Syncing content
08:36:07 [content] Synced content
08:36:07 [vite] Re-optimizing dependencies because lockfile has changed

 astro  v0.0.0-dev-after-sync-20241223135326 ready in 435 ms

┃ Local    http://localhost:4321/
┃ Network  use --host to expose

▶ This is a  dev-after-sync-20241223135326  prerelease build!
  Report issues here: https://astro.build/issues

08:36:07 watching for file changes...
08:36:14 [200] / 13ms
The collection "blog" does not exist or is empty. Please check your content config file for errors.
08:36:16 [200] /blog 9ms
```
</details>

**Comment by ericswpark** [comment]
@ascorbic yes, with the reproduction steps above I was able to reproduce it in the minimal repro repo as well.

**Comment by Gazook89** [comment]
@ericswpark 

> Once you reproduce using those steps, your .astro/ folder should be missing the data-store.json file.

Should i be expecting to see a `data-store.json` file in that folder if everything is working correctly?  I have never seen it, so I'm just curious.  It'd be easier to spot that than switching back to my browser tab to check if something has worked.

**Comment by ericswpark** [comment]
@Gazook89 if you've never seen it, probably not. I just did a `diff` between when the collection rendered properly and when it wouldn't. I think the only thing you can count on is the warning message: `The collection "<collection name>" does not exist or is empty. Please check your content config file for errors.`

**Comment by Gazook89** [comment]
I don't know if issues can be reopened in this repo, or if I should create a new one, but the v5.1.3 update to which #12818 was attached to hasn't resolved this issue for me.

To be sure, I created a new repo, did `npm create astro@latest`, set it up as the Blog template, ran `npm run dev` and as soon as I went to the `/blog` route got the same error:

```
16:24:49 [types] Generated 0ms
16:24:49 [content] Syncing content
16:24:50 [content] Synced content
16:24:50 [vite] Re-optimizing dependencies because vite config has changed

 astro  v5.1.3 ready in 414 ms

┃ Local    http://localhost:4321/
┃ Network  use --host to expose

16:24:50 watching for file changes...
The collection "blog" does not exist or is empty. Please check your content config file for errors.
16:24:54 [200] /blog 14ms"
```

I changed nothing from the template.  I reran `npm i` just to be sure.  As before, `build` and `preview` works fine, but obviously is a pain during development (at least compared to a working `dev` server).

Using the v5.1.3 with the Blog template on Stackblitz works fine.

```
Astro                    v5.1.3
Node                     v22.12.0
System                   macOS (arm64)
Package Manager          npm
Output                   static
Adapter                  none
Integrations             @astrojs/mdx
                         @astrojs/sitemap
```

**Comment by ascorbic** [comment]
@Gazook89 I'm going to need a repro I'm afraid. I followed your steps and it works fine for me. This is what I did:

```
❯ npm create astro@latest

> npx
> create-astro


 astro   Launch sequence initiated.

   dir   Where should we create your new project?
         ./tasty-telescope

  tmpl   How would you like to start your new project?
         Use blog template

  deps   Install dependencies?
         Yes

   git   Initialize a new git repository?
         Yes

      ✔  Project initialized!
         ■ Template copied
         ■ Dependencies installed
         ■ Git initialized

  next   Liftoff confirmed. Explore your project!

         Enter your project directory using cd ./tasty-telescope
         Run npm run dev to start the dev server. CTRL+C to stop.
         Add frameworks like react or tailwind using astro add.

         Stuck? Join us at https://astro.build/chat

╭─────╮  Houston:
│ ◠ ◡ ◠  Good luck out there, astronaut! 🚀
╰─────╯
❯ cd tasty-telescope
❯ npm dev
Unknown command: "dev"


Did you mean this?
  npm run dev # run the "dev" package script
To see a list of supported npm commands, run:
  npm help
❯ npm run dev

> tasty-telescope@0.0.1 dev
> astro dev

11:24:08 [types] Generated 0ms
11:24:09 [content] Syncing content
11:24:09 [content] Synced content

 astro  v5.1.3 ready in 403 ms

┃ Local    http://localhost:4321/
┃ Network  use --host to expose

11:24:09 watching for file changes...
11:24:24 [200] / 11ms
```

Can you confirm if there's anythign that you did different?

**Comment by ascorbic** [comment]
OK, this is interesting. This doesn't occur with pnpm - just npm. It also works as soon as you first save a file.

**Comment by Gazook89** [comment]
> It also works as soon as you first save a file.

This isn't my experience as of a few minutes ago, as of Astro 5.1.3.  I just wiped my test blog (created with the Astro Blog Template) and installed a new one.  On NPM, if I do `npm run dev`, and then modify a file and save, I'm still hitting the empty collection problem.  

So I'm waiting for this new patch to come down the pipe to try that out, but just noting here that right now, saving a file doesn't fix the problem (for me).

**Comment by Gazook89** [comment]
I just updated my test blog, which is the fresh install of 5.1.3 with Blog template, to 5.1.4 and still have the issue.  I also tried `npm run astro sync` just to try something and no luck.  I also tried creating a new post after starting the dev server and saving it, but it didn't change anything.

I'm comforted to see that another user mentioned the same issue with a fresh install in #12939, so I know it's not just some dumb config on my computer.

**Comment by ascorbic** [comment]
@Gazook89 the example in #12939 is still using 5.1.3. Can you share a new reproduction, because I can't reproduce it in 5.1.4 using the previous steps.

**Comment by Gazook89** [comment]
In doing this, I updated to **5.1.5** and it seems to work on test blog I setup for this.  I won't have a chance to test on my actual blog until later today, but this is an improvement on the test blog and I don't see why it wouldn't apply to the real one.  

So I think we can say it's resolved now with fingers crossed and I'll come back here if for some reason it isn't.

Thanks @ascorbic for your persistence, this is all started in an effort to use your read-csv plugin :)
